crypto/sm2/sm2_pmeth.c

```diff
@@ -0,0 +1,282 @@
+/*
+ * Copyright 2006-2018 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the OpenSSL license (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include "internal/cryptlib.h"
+#include <openssl/asn1t.h>
+#include <openssl/ec.h>
+#include <openssl/evp.h>
+#include "internal/evp_int.h"
+#include "internal/sm2.h"
+#include "internal/sm2err.h"
+
+/* EC pkey context structure */
+
+typedef struct {
+    /* Key and paramgen group */
+    EC_GROUP *gen_group;
+    /* message digest */
+    const EVP_MD *md;
+} SM2_PKEY_CTX;
+
+static int pkey_sm2_init(EVP_PKEY_CTX *ctx)
+{
+    SM2_PKEY_CTX *dctx;
+
+    if ((dctx = OPENSSL_zalloc(sizeof(*dctx))) == NULL) {
+        SM2err(SM2_F_PKEY_SM2_INIT, ERR_R_MALLOC_FAILURE);
+        return 0;
+    }
+
+    ctx->data = dctx;
+    return 1;
+}
+
+static int pkey_sm2_copy(EVP_PKEY_CTX *dst, EVP_PKEY_CTX *src)
+{
+    SM2_PKEY_CTX *dctx, *sctx;
+
+    if (!pkey_sm2_init(dst))
+        return 0;
+    sctx = src->data;
+    dctx = dst->data;
+    if (sctx->gen_group != NULL) {
+        dctx->gen_group = EC_GROUP_dup(sctx->gen_group);
+        if (dctx->gen_group != NULL)
+            return 0;
+    }
+    dctx->md = sctx->md;
+
+    return 1;
+}
+
+static void pkey_sm2_cleanup(EVP_PKEY_CTX *ctx)
+{
+    SM2_PKEY_CTX *dctx = ctx->data;
+
+    if (dctx != NULL) {
+        EC_GROUP_free(dctx->gen_group);
+        OPENSSL_free(dctx);
+    }
+}
+
+static int pkey_sm2_sign(EVP_PKEY_CTX *ctx, unsigned char *sig, size_t *siglen,
+                        const unsigned char *tbs, size_t tbslen)
+{
+    int ret;
+    unsigned int sltmp;
+    EC_KEY *ec = ctx->pkey->pkey.ec;
+
+    if (sig == NULL) {
+        *siglen = ECDSA_size(ec);
+        return 1;
+    }
+
+    if (*siglen < (size_t)ECDSA_size(ec)) {
```

Comment:
**Formally speaking** this is problematic cast.  Because if ECDSA_size returns negative value, then this condition would be met. Well, same can be said about above call to ECDSA_size. Two options, a) leave comment in the beginning that we count on ECDSA_size to return small positive (one can express it in form of "soft" assert); b) explicitly check for ECDSA_size return value being positive in the beginning of function and return failure if it's not. In either case it would be appropriate to add local variable that accepts ECDSA_size value.

Reply:
I will fix here, same logic is used in ec_pmeth.c btw